src/utils/yaml.ts

```diff
@@ -324,10 +326,45 @@ export function isValueEscapedAlready(value: string): boolean {
  * @param {string} value The value to escape if possible
  * @param {string} defaultEscapeCharacter The character escape to use around the value if a specific escape character is not needed.
  * @param {boolean} forceEscape Whether or not to force the escaping of the value provided.
+ * @param {boolean} skipValidation Whether or not to ensure that the result string could be unescaped back to the value.
  * @return {string} The escaped value if it is either necessary or forced and the provided value if it cannot be escaped, is escaped,
  * or does not need escaping and the force escape is not used.
  */
-export function escapeStringIfNecessaryAndPossible(value: string, defaultEscapeCharacter: string, forceEscape: boolean = false): string {
+export function escapeStringIfNecessaryAndPossible(value: string, defaultEscapeCharacter: QuoteCharacter, forceEscape: boolean = false, skipValidation: boolean = false): string {
+  const basicEscape = basicEscapeString(value, defaultEscapeCharacter, forceEscape);
+  if (skipValidation) {
+    return basicEscape;
+  }
+
+  try {
+    const unescaped = load(basicEscape) as string;
+    if (unescaped === value) {
+      return basicEscape;
+    }
+  } catch {
+    // invalid YAML
+  }
+
+  const escapeWithDefaultCharacter = dump(value, {
+    lineWidth: -1,
+    quotingType: defaultEscapeCharacter,
+    forceQuotes: forceEscape,
+  }).slice(0, -1);
+
+  const escapeWithOtherCharacter = dump(value, {
+    lineWidth: -1,
+    quotingType: defaultEscapeCharacter == '"' ? '\'' : '"',
+    forceQuotes: forceEscape,
+  }).slice(0, -1);
+
+  if (escapeWithOtherCharacter === value || escapeWithOtherCharacter.length < escapeWithDefaultCharacter.length) {
+    return escapeWithOtherCharacter;
+  }
```

Comment:
Could you explain how these checks determine that you should use `escapeWithOtherCharacter`?

My confusion may come from a lack of understanding around `dump`, but it looks to me like the string could need escaping with a double quote (not a single quote) and this would fail to yield the proper result if the default character was `'`. However it may be that `js-yaml` does some weird escaping where it uses `'` to somehow escape a single quote character.

Reply:
@pjkaufman 

I checked the way `dump` works

You call `result = dump(x, { quotingType: y })`

| x   | y   | result |
|-----|-----|--------|
| `"` | `"` | `"\""` |
| `'` | `"` | `"'"`  |
| `"` | `'` | `'"'`  |
| `'` | `'` | `''''` |

I ran all the results above through YAML validator and confirmed they are all valid YAML strings and represent the value in `x`

The idea in my implementation was to see which of the quotes characters provide shorter output and use it.

For the synthetic examples in the table above, we see that when we escape single quote character `'` (lines 2 and 4), it's better to escape with a double quote character `"` (line 2). When we escape double quote character `"` (lines 1 and 3), it's better to escape with a single quote characer `'` (line 3). That matches the logic in the previous implementation